reV/offshore/offshore.py

```diff
@@ -165,75 +169,172 @@ def _parse_cf_meta(gen_fpath):
+        offshore_data = offshore_data[mask]
+        offshore_data = offshore_data.sort_values('gid')
+        offshore_data = offshore_data.reset_index(drop=True)
+
+        return offshore_data
+
+    def _parse_system_inputs(self):
+        """Get the system inputs dict (SAM tech inputs) from project points.
 
         Returns
         -------
-        d : np.ndarray
-            Distance between offshore resource pixel and offshore wind farm.
-        i : np.ndarray
-            Offshore farm row numbers corresponding to resource pixels
-            (length is number of offshore resource pixels in gen_fpath).
-        d_lim : float
-            Maximum distance limit between wind farm points and resouce pixels.
+        system_inputs : pd.DataFrame
+            DataFrame of SAM config inputs (columns) for every offshore
+            resource gid (row). Index is resource gids and there is also
+            a column "gid" with the copied gids.
         """
-        # pylint: disable=not-callable
-        tree = cKDTree(self._farm_coords)
-        lat_lon_cols = get_lat_lon_cols(self.meta_source_offshore)
-        d, i = tree.query(self.meta_source_offshore[lat_lon_cols])
 
-        d_lim = 0
-        if len(self._farm_coords) > 1:
-            d_lim, _ = tree.query(self._farm_coords, k=2)
-            d_lim = 0.5 * np.median(d_lim[:, 1])
-            i[(d > d_lim)] = -1
+        system_inputs = {}
+
+        for gid in self.offshore_res_gids:
+            system_inputs[gid] = self._project_points[gid][1]
 
-        return d, i, d_lim
+            if 'turbine_capacity' not in system_inputs[gid]:
+                # convert from SAM kw powercurve to MW.
+                arr = system_inputs[gid]['wind_turbine_powercurve_powerout']
+                cap_kw = np.max(arr)
+                cap_mw = cap_kw / 1000
+                system_inputs[gid]['turbine_capacity'] = cap_mw
+
+        system_inputs = pd.DataFrame(system_inputs).T
+        system_inputs = system_inputs.sort_index()
+        system_inputs['gid'] = system_inputs.index.values
+        system_inputs.index.name = 'gid'
+
+        return system_inputs
+
+    def _preflight_checks(self):
+        """Run some preflight checks on the offshore inputs"""
+        sam_configs = {k: v for k, v in
+                       self._project_points.sam_inputs.items()
+                       if k in self._nrwal_configs}
+        for cid, sys_in in sam_configs.items():
+            if 'turbine_capacity' not in sys_in:
+                msg = ('System input key "turbine_capacity" not found in '
+                       'SAM system inputs for "{}". Calculating from turbine '
+                       'power curves.'.format(cid))
+                logger.warning(msg)
+                warn(msg, OffshoreWindInputWarning)
+
+            loss1 = sys_in.get('wind_farm_losses_percent', 0)
+            loss2 = sys_in.get('turb_generic_loss', 0)
+            if loss1 != 0 or loss2 != 0:
```

Comment:
since the loses are a scalar hair-cut why not raise the warning and remove the losses that were applied?

Reply:
we could do that.... I'm not sure its worth the complexity and the implicit manipulation of the data. There are also a ton of different loss terms now in the sam turbine configs that we would have to check for and combine.... I'm not crazy about the idea.....

Reply:
You sold me at a ton of new loss terms...